src/experimental/job/job.ts

```diff
@@ -139,17 +139,12 @@ export class Job<Output = unknown> {
       const exeUnit = await rental.getExeUnit();
       this.events.emit("started");
 
-      const onAbort = async () => {
-        await rental.stopAndFinalize();
-        this.events.emit("canceled");
-      };
-
       if (signal.aborted) {
-        await onAbort();
+        this.events.emit("canceled");
         throw new GolemAbortError("Canceled");
       }
 
-      signal.addEventListener("abort", onAbort, { once: true });
+      signal.addEventListener("abort", () => this.events.emit("canceled"), { once: true });
```

Comment:
The issue now is that when the work is already started, aborting the job will **not** stop it. The unit test fails to catch this because the signal gets aborted before the actual work begins, you can see this if you modify the test to use a `jest fn`:
```ts
const taskFn = jest.fn().mockImplementation(() => {
  return new Promise((resolve) => setTimeout(resolve, 10000));
});
job.startWork(taskFn);

// flush promises
// await new Promise(setImmediate); // <------------- uncomment to cause a timeout
await job.cancel();
await expect(job.waitForResult()).rejects.toThrow("Canceled");

verify(mockRental.stopAndFinalize()).once();
expect(taskFn).toHaveBeenCalled(); // <---------------- error! 0 calls
```
(the abort error is thrown on line 144)
If you uncomment the `Promise(setImmediate)` then the test will time out, because the task function will start and aborting it won't do anything.

To fix this, I think we should race the `workOnGolem` promise with the abort signal, something like
```ts
const workPromise = workOnGolem(exeUnit);
const abortPromise = new Promise<never>((resolve, reject) => {
  signal.addEventListener(
    "abort",
    () => {
      this.events.emit("canceled");
      reject(new GolemAbortError("Canceled"));
    },
    { once: true },
  );
});
return Promise.race([workPromise, abortPromise]);
```

Reply:
I'm not sure if this is needed. All we need to do is pass the `signal` to `getExeUnit` method and then it will be passed to the activity executor and each attempt to do something or fetch the result will end with a `GolemAbortError`, which will stop the work.

Reply:
You're right, I didn't think of that